### PR TITLE
[MRG+2] refactor linear_regression_raw

### PR DESCRIPTION
**Pull request by jona-sassenhagen**
Factor out a bunch of helpers. Prerequisite for 
- encoding models
- reducing redundancy with XDawn (#2332)

### CODE REVIEW

**Comment by Eric89GXL** [comment]
LGTM

File: mne/stats/tests/test_regression.py [hunk 1]
```diff
@@ -91,6 +91,7 @@ def test_continuous_regression_no_overlap():
     tmin, tmax = -.1, .5
 
     raw = mne.io.Raw(raw_fname, preload=True)
+    raw.apply_proj()
```
Comment (agramfort) [inline mne/stats/tests/test_regression.py]: why?

Comment (jona-sassenhagen) [inline mne/stats/tests/test_regression.py]: We compare it to the regular evoked object and that one seemingly gets a projection applied (?). So this makes the objects we compare more similar.

File: mne/stats/regression.py [hunk 2]
```diff
@@ -256,7 +256,39 @@ def solver(X, Y):
...
+    # build predictors
+    X, conds, cond_length, tmin_s, tmax_s = _prepare_rerp_preds(
+        n_samples=data.shape[1], sfreq=info["sfreq"], events=events,
+        event_id=event_id, tmin=tmin, tmax=tmax, covariates=covariates)
+
+    # remove "empty" and contaminated data points
+    X, data = _clean_rerp_input(X, data, reject, flat, decim, info, tstep)
+
+    # solve linear system
+    coefs = solver(X, data)
```
Comment (kingjr) [inline mne/stats/regression.py]: IMO, we should avoid `data` as a var name. What is this actually, a regressor? maybe `y`?

File: mne/stats/tests/test_regression.py (hunk 1 above)
Comment (agramfort) [inline mne/stats/tests/test_regression.py]: Does it allow to reduce a tol on a subsequent test?

File: mne/stats/regression.py (hunk 2 above)
Comment (dengemann) [inline mne/stats/regression.py]: yes, if we know more than that it's data let's indicate that knowledge by choosing a more specific name.

Comment (jona-sassenhagen) [inline mne/stats/regression.py]: In principle fine with me, but I think this one I actually settled on with @agramfort , or at least he sat next to me when I wrote that part, maybe he has a specific reason.

File: mne/stats/tests/test_regression.py (hunk 1 above)
Comment (jona-sassenhagen) [inline mne/stats/tests/test_regression.py]: Will check, hopefully.

**Comment by jona-sassenhagen** [comment]
I think I want to factor out contructing the evokeds too, cause that makes it fairly easy to actually plugin in sklearn objects here.

And I'd like to add an example for how to combine this and sklearn for a really easy encoder model validation.

Sorry for going back and forth on MRG/WIP all the time :)

File: mne/stats/tests/test_regression.py (hunk 1 above)
Comment (jona-sassenhagen) [inline mne/stats/tests/test_regression.py]: Yes, it does :)

**Comment by jona-sassenhagen** [comment]
Ready now.

**Comment by Eric89GXL** [comment]
+1 for merge

File: mne/stats/regression.py (hunk 2 above)
Comment (dengemann) [inline mne/stats/regression.py]: Let's see if this is worth a separate issue, should be systematic then across the package. Let's not touch it here.

**Comment by jona-sassenhagen** [comment]
The 2nd Travis build is stuck ...

**Comment by Eric89GXL** [comment]
Restarted it for you, will merge if it comes back happy

**Comment by Eric89GXL** [comment]
Thanks @jona-sassenhagen

**Comment by jona-sassenhagen** [comment]
Thanks for reviews everyone!

**Pull request merged**